### PR TITLE
fix: Upgrade to SageMaker v2 SDK

### PR DESCRIPTION
**Pull request by brightsparc**
*Issue #, if available:* #69 Feature request: support SageMaker SDK 2.0

Related to #73 

*Description of changes:*

Manually upgrade the step functions SDK to support the [breaking changes](https://sagemaker.readthedocs.io/en/stable/v2.html#breaking-changes) relevant to this library specifically.

* `image` -> `image_uri`
* `train_instance_count` -> `instance_count`
* `train_instance_type` -> `instance_type`
* `train_max_run` -> `max_run`
* `train_max_run_wait`  -> `max_run_wait`
* `train_volume_size`  ->  `volume_size`
* `sagemaker.session.s3_input` -> `sagemaker.inputs.TrainingInput`

I upgrade dependencies to:
```
sagemaker>=2.1.0`
boto3>=1.14.38
```
I upgrade package version to:  `2.0.0`

I have an integration tests that validates this code works with - but have a few issues running the complete test suite.
* TrainingStep
* ModelStep
* ProcessingStep

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by brightsparc** [comment]
I have validate that there are no more upgrades required by the tool after performing upgrade-v2 to all files and using [black](https://black.readthedocs.io/en/stable/installation_and_usage.html) to format consistency given the upgrade tool messes with formatting.  (although I haven't re-formatted the code using black as part of this branch).

```
upgrade_dir () {    
    shopt -s nullglob dotglob

    for pathname in "$1"/*; do
        if [ -d "$pathname" ]; then
            upgrade_dir "$pathname"
        else
            case "$pathname" in
                *.py)
                    sagemaker-upgrade-v2 --in-file "$pathname" --out-file "$pathname"
            esac
        fi
    done
}

# install black for formatting
pip install black

# apply upgrades
upgrade_dir ./src
upgrade_dir ./tests

# re-apply black to src and test
black -S ./src
black -S ./tests

git diff -w
```

**Comment by vaib-amz** [comment]
Thanks a lot for this PR @brightsparc ! I don't have any feedback besides the docstring. I will ask another team member to look at this PR, and will also have to get a confirmation regarding the version bump.

**Comment by pouyanhoss** [comment]
@brightsparc 
Hi Julian,

I am using DeepAR in SageMaker and recently upgraded it to SDK v2. I was able to modify everything except the following part:

`predictor = DeepARPredictor(endpoint=best_tuning_job_name,
                            sagemaker_session=sess,
                            content_type='application/json')`

Based on the documentation, I modified it as following:

`predictor = DeepARPredictor(endpoint_name=best_tuning_job_name,
sagemaker_session=sess,serializer=sagemaker.serializers.JSONSerializer())`

However, I get a TypeError when I call the predictor in the next cell: 

`predictor.set_prediction_parameters(freq, prediction_length)
list_of_df = predictor.predict(violation_list_training[:2])`

**TypeError: Object of type bytes is not JSON serializable**

Any help would be much appreciated.

**Review by yoodan93 — DISMISSED** [review]
Looks good to me

**Comment by wong-a** [comment]
Thanks for the contribution @brightsparc 

A few things to take care of before merging:
1. Resolve merge conflicts in VERSION, requiremenets.txt, and setup.py
1. Make sure acceptance and unit tests pass

For the Step Functions team, we need to figure out the plan for maintaining v1 of the stepfunctions SDK and user guidance for migration from v1 to v2.

File: src/stepfunctions/steps/sagemaker.py [hunk 1]
```diff
@@ -202,7 +202,7 @@ def __init__(self, state_id, model, model_name=None, instance_type=None, tags=No
             tags (list[dict], optional): `List to tags <https://docs.aws.amazon.com/sagemaker/latest/dg/API_Tag.html>`_ to associate with the resource.
```
[the remark this answers is not in the document]
Comment (brightsparc) [reply src/stepfunctions/steps/sagemaker.py:202]: Update the branch to fix docstring as per comment.

**Comment by brightsparc** [comment]
> For the Step Functions team, we need to figure out the plan for maintaining v1 of the stepfunctions SDK and user guidance for migration from v1 to v2.

I have resolved the conflict to stick with the v2 target version.  @wong-a I can revise this as required.

File: src/stepfunctions/steps/sagemaker.py [hunk 2]
```diff
@@ -36,12 +36,12 @@ def __init__(self, state_id, estimator, job_name, data=None, hyperparameters=Non
             data: Information about the training data. Please refer to the ``fit()`` method of the associated estimator, as this can take any of the following forms:
 
                 * (str) - The S3 location where training data is saved.
-                * (dict[str, str] or dict[str, sagemaker.session.s3_input]) - If using multiple
+                * (dict[str, str] or dict[str, sagemaker.inputs.TrainingInput]) - If using multiple
```
Comment (metrizable) [inline src/stepfunctions/steps/sagemaker.py:39]: suggested edit: is it worth considering for the future is to specify the generic [`Dict`](https://docs.python.org/3/library/typing.html#typing.Dict) from typing (or even [`Mapping`](https://docs.python.org/3/library/typing.html#typing.Mapping))? I know these are just docstring comments, but in the future, you may want to include type hinting and the change will be minor in the future at that point:

```suggestion
                * (Dict[str, str] or Dict[str, sagemaker.inputs.TrainingInput]) - If using multiple
```

**Review by metrizable — APPROVED** [review]
looks good! is there interest in using an auto-formatter like [`black`](https://github.com/psf/black)?

File: src/stepfunctions/steps/sagemaker.py (hunk 2 above)
Comment (brightsparc) [reply src/stepfunctions/steps/sagemaker.py:39]: I think an auto-formatter is a good idea.  I used `black` when I was ensuring I had applied the correct v2 SM upgrades, but the diff made it hard to compare changes so left it as is.

Those doc comments could be updated with a seperate PR as I suspect there are a few.

**Comment by vaib-amz** [comment]
Updating version to a pre-release candidate. The tests were all okay, only the version has changed. Since the changed were approved already, merging in.

**Pull request merged**